### PR TITLE
Implement Prism -> Sorbet translation for "implicit rest" nodes

### LINKED ISSUE #105: Implement translation for `PM_IMPLICIT_REST_NODE`

### PR DESCRIPTION
**Pull request by amomchilov**
Closes #105

### CODE REVIEW

File: parser/prism/Translator.cc [hunk 1]
```diff
@@ -757,58 +761,15 @@ unique_ptr<parser::Node> Translator::translate(pm_node_t *node) {
...
-
-            NodeVec sorbetExpressions{};
-            sorbetExpressions.reserve(prismLefts.size() + prismRights.size() + (prismRestNode != nullptr ? 1 : 0));
-
-            translateMultiInto(sorbetExpressions, prismLefts);
-
-            if (prismRestNode != nullptr) {
-                // We don't just call `translate()` because that would create a `parser::Splat`, but we need
-                // `parser::SplatLhs`.
-                auto splatNode = down_cast<pm_splat_node>(prismRestNode);
```
Comment (amomchilov) [inline parser/prism/Translator.cc:772]: This was an incorrect assumption, and used to be unchecked prior to #302. This variable can now be a `PM_IMPLICIT_REST_NODE`, and is handled correctly.

https://github.com/Shopify/sorbet/blob/ee541d10ae57b2bc471771c054fbd689eb6ba643/parser/prism/Translator.cc#L799-L802

File: parser/prism/Translator.cc [hunk 2]
```diff
@@ -1638,4 +1598,47 @@ unique_ptr<parser::Regexp> Translator::translateRegexp(pm_string_t unescaped, co
...
 
+// Creates a `parser::Mlhs` for either a `PM_MULTI_WRITE_NODE` or `PM_MULTI_TARGET_NODE`.
+template <typename PrismNode> std::unique_ptr<parser::Mlhs> Translator::translateMultiTargetLhs(PrismNode *node) {
+    static_assert(
+        is_same_v<PrismNode, pm_multi_target_node> || is_same_v<PrismNode, pm_multi_write_node>,
+        "Translator::translateMultiTarget can only be used for PM_MULTI_TARGET_NODE and PM_MULTI_WRITE_NODE.");
+
+    auto location = translateLoc(node->base.location);
+
+    // Left-hand side of the assignment
```
Comment (amomchilov) [inline parser/prism/Translator.cc:1609]: I suppose this comment doesn't make much sense here anymore

File: parser/prism/Translator.cc [hunk 3]
```diff
@@ -1638,4 +1598,47 @@ unique_ptr<parser::Regexp> Translator::translateRegexp(pm_string_t unescaped, co
...
+    auto prismRights = absl::MakeSpan(node->rights.nodes, node->rights.size);
+    auto prismSplat = node->rest;
+
+    NodeVec sorbetLhs{};
+    sorbetLhs.reserve(prismLefts.size() + prismRights.size() + (prismSplat != nullptr ? 1 : 0));
+
+    translateMultiInto(sorbetLhs, prismLefts);
+
+    if (prismSplat != nullptr) {
+        switch (PM_NODE_TYPE(prismSplat)) {
```
Comment (amomchilov) [inline parser/prism/Translator.cc:1620]: I wish I could do without this nullcheck, and just be able to handle in another `case`. Opened https://github.com/ruby/prism/issues/3189 to discuss

File: parser/prism/Helpers.h [hunk 4]
```diff
@@ -185,11 +185,9 @@ template <typename PrismNode> PrismNode *down_cast(pm_node_t *anyNode) {
...
                   "The `down_cast` function should only be called on Prism node pointers.");
     ENFORCE(PM_NODE_TYPE_P(anyNode, PrismNodeTypeHelper<PrismNode>::TypeID),
-        "Failed to cast a Prism AST Node. Expected {} (#{}), but got {} (#{}).",
-            pm_node_type_to_str(PrismNodeTypeHelper<PrismNode>::TypeID),
-            PrismNodeTypeHelper<PrismNode>::TypeID,
-            pm_node_type_to_str(PM_NODE_TYPE(anyNode)),
-            PM_NODE_TYPE(anyNode));
+            "Failed to cast a Prism AST Node. Expected {} (#{}), but got {} (#{}).",
+            pm_node_type_to_str(PrismNodeTypeHelper<PrismNode>::TypeID), PrismNodeTypeHelper<PrismNode>::TypeID,
+            pm_node_type_to_str(PM_NODE_TYPE(anyNode)), PM_NODE_TYPE(anyNode));
```
Comment (egiurleo) [inline parser/prism/Helpers.h:190]: I feel like there are a lot of formatting changes in recent PRs -- what's up with that?

**Review by egiurleo — APPROVED** [review]
Woohoo! Love the de-duplication.

**Pull request merged**